pennylane/devices/qutrit_mixed/apply_operation.py

```diff
@@ -0,0 +1,276 @@
+    else:
+        _state = _conjugate_state_with(op.matrix())
+
+    # Permute the affected axes to their destination places.
+    # The row indices of the kraus operators are moved from the beginning to the original
+    # target row locations, the column indices from the end to the target column locations
+    source_left = list(range(num_ch_wires))
+    dest_left = row_wires_list
+    source_right = list(range(-num_ch_wires, 0))
+    dest_right = col_wires_list
+
+    print(source_left + source_right)
+    print(dest_left + dest_right)
+    return math.moveaxis(_state, source_left + source_right, dest_left + dest_right)
+
+
+@singledispatch
+def apply_operation(
+    op: qml.operation.Operator, state, is_state_batched: bool = False, debugger=None
+):
+    """Apply and operator to a given state.
+
+    Args:
+        op (Operator): The operation to apply to ``state``
+        state (TensorLike): The starting state.
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+        debugger (_Debugger): The debugger to use
+
+    Returns:
+        ndarray: output state
+
+    .. warning::
+
+        ``apply_operation`` is an internal function, and thus subject to change without a deprecation cycle.
+
+    .. warning::
+        ``apply_operation`` applies no validation to its inputs.
+
+        This function assumes that the wires of the operator correspond to indices
+        of the state. See :func:`~.map_wires` to convert operations to integer wire labels.
+
+        The shape of state should be ``[qudit_dim]*(num_wires * 2)``, where ``qudit_dim`` is
+        the dimension of the system.
+
+    This is a ``functools.singledispatch`` function, so additional specialized kernels
+    for specific operations can be registered like:
+
+    .. code-block:: python
+
+        @apply_operation.register
+        def _(op: type_op, state):
+            # custom op application method here
+
+    **Example:**
+
+    >>> state = np.zeros((3,3))
+    >>> state[0][0] = 1
+    >>> state
+    tensor([[1., 0., 0.],
+        [0., 0., 0.],
+        [0., 0., 0.]], requires_grad=True)
+    >>> apply_operation(qml.TShift(0), state)
+    tensor([[0., 0., 0.],
+        [0., 1., 0],
+        [0., 0., 0.],], requires_grad=True)
+
+    """
+    return _apply_operation_default(op, state, is_state_batched, debugger)
+
+
+def _apply_operation_default(op, state, is_state_batched, debugger):
+    """The default behaviour of apply_operation, accessed through the standard dispatch
+    of apply_operation, as well as conditionally in other dispatches.
+    """
+    if (
+        len(op.wires) < EINSUM_OP_WIRECOUNT_PERF_THRESHOLD
+        and math.ndim(state) < EINSUM_STATE_WIRECOUNT_PERF_THRESHOLD
+    ) or (op.batch_size or is_state_batched):
+        return apply_operation_einsum(op, state, is_state_batched=is_state_batched)
+    # TODO fix state batching on tensordot
```

Comment:
Right now state batching doesn't work for tensordot, it could but I wanted to put this up and it is not necessary other than for performance in some cases as it can be handled by einsum. I could add state batching for tensordot now if it is needed.